File: src/service/AuthService.js
const jwt = require('jsonwebtoken');
const bcrypt = require('bcryptjs');
const config = require('../config/app');
const User = require('../model/User');
const Token = require('../model/Token');

class AuthService {
  static async loginAdmin(username, password) {
    const user = await User.findByUsername(username);

    if (!user) {
      throw new Error('Username tidak ditemukan');
    }

    if (user.role !== 'superadmin') {
      throw new Error('Akses ditolak');
    }

    const match = await bcrypt.compare(password, user.password);
    if (!match) {
      throw new Error('Password salah');
    }

    const token = jwt.sign(
      { role: 'superadmin', type: 'admin', userId: user.id },
      config.jwt.secret,
      { expiresIn: config.jwt.expiresIn }
    );

    return { token, role: 'superadmin' };
  }

  static async loginWithToken(accessCode) {
    const tokenData = await Token.findByCode(accessCode);

    if (!tokenData) {
      throw new Error('Kode Akses tidak valid');
    }

    const allowedApps = typeof tokenData.allowed_apps === 'string'
      ? JSON.parse(tokenData.allowed_apps)
      : tokenData.allowed_apps;

    const token = jwt.sign(
      {
        role: 'user',
        type: 'token',
        token_id: tokenData.id,
        allowedApps,
      },
      config.jwt.secret,
      { expiresIn: '12h' }
    );

    return {
      token,
      role: 'user',
      label: tokenData.label,
      allowedApps,
    };
  }

  static verifyToken(token) {
    try {
      return jwt.verify(token, config.jwt.secret);
    } catch (error) {
      if (error.name === 'TokenExpiredError') {
        throw new Error('Sesi kadaluarsa');
      }
      throw new Error('Token tidak valid');
    }
  }
}

module.exports = AuthService;

File: internal/service/auth_service.go
package service

import (
	"context"
	"errors"
	"time"

	"golang.org/x/crypto/bcrypt"

	"github.com/tragasolusi/pm2-manager-api/internal/auth"
	"github.com/tragasolusi/pm2-manager-api/internal/config"
	"github.com/tragasolusi/pm2-manager-api/internal/repository"
)

var (
	ErrInvalidCredentials = errors.New("username atau password salah")
	ErrInvalidAccessCode  = errors.New("kode akses tidak valid")
	ErrAccessDenied       = errors.New("akses ditolak")
)

type AuthService struct {
	users  *repository.UserRepository
	tokens *repository.TokenRepository
	jwt    *auth.Service
	cfg    *config.AppConfig
}

func NewAuthService(
	users *repository.UserRepository,
	tokens *repository.TokenRepository,
	jwt *auth.Service,
	cfg *config.AppConfig,
) *AuthService {
	return &AuthService{users: users, tokens: tokens, jwt: jwt, cfg: cfg}
}

type LoginResult struct {
	Token       string   `json:"token"`
	Role        string   `json:"role"`
	Label       string   `json:"label,omitempty"`
	AllowedApps []string `json:"allowedApps,omitempty"`
}

func (s *AuthService) LoginAdmin(ctx context.Context, username, password string) (*LoginResult, error) {
	u, err := s.users.FindByUsername(ctx, username)
	if err != nil {
		return nil, ErrInvalidCredentials
	}
	if u.Role != "superadmin" {
		return nil, ErrAccessDenied
	}
	if err := bcrypt.CompareHashAndPassword([]byte(u.Password), []byte(password)); err != nil {
		return nil, ErrInvalidCredentials
	}
	tok, err := s.jwt.IssueAdmin(u.ID)
	if err != nil {
		return nil, err
	}
	return &LoginResult{Token: tok, Role: "superadmin"}, nil
}

func (s *AuthService) LoginWithToken(ctx context.Context, accessCode string) (*LoginResult, error) {
	t, err := s.tokens.FindByCode(ctx, accessCode)
	if err != nil {
		return nil, ErrInvalidAccessCode
	}
	tok, err := s.jwt.IssueTokenUser(t.ID, t.Label, t.AllowedApps, 12*time.Hour)
	if err != nil {
		return nil, err
	}
	return &LoginResult{
		Token:       tok,
		Role:        "user",
		Label:       t.Label,
		AllowedApps: t.AllowedApps,
	}, nil
}
